### ISSUE TITLE
two optimizers in MGDA

### ISSUE THREAD

**Issue by turboxin** [pxiangwu/MotionNet#15]
Hi @pxiangwu ,

First of all, thanks for opensourcing this great work!

I notice that there are two optimizers in MGDA training, one for shared feature extraction and one for heads. Could you kindly explain why is this necessary. In the origin implemantation, they seem to use only one optimizer [https://github.com/isl-org/MultiObjectiveOptimization/blob/master/multi_task/train_multi_task.py](url)  .

Looking forward for your reply. Thanks  a lot!

**Comment by pxiangwu** [comment]
Hello,

For your question, did you mean the MGDA in this line of MotionNet code: [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L362)? Actually the MGDA is applied to the **three** heads and not employed to optimize the shared feature extraction backbone. See the code [here](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L310), [here](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L331), and [here](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L356). In other words, we utilized MGDA to compute the loss coefficients (aka, balancing factors) for the three heads. With the computed loss coefficients, we then apply gradient descent optimizer (e.g., Adam) to train the network as usual.

Hope the above explanations answer your question.

**Comment by turboxin** [comment]
Hi @pxiangwu ,

Thanks for your really quick reply. I understand what you're talking about, that the MGDA is applied to the three heads and not employed to optimize the shared feature extraction backbone. 

I actually mean the lines [here](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L305) and [here](https://github.com/eniolasonowo/MotionNet/blob/ab664d594eef6e2aa88fd8f165c04086794413c1/train_single_seq_MGDA.py#L306), where the two optimizers are first zero-graded and then used to compute the grad. In mine understanding it seems that one optimizer could do the same thing, which is the case for origin implemantation of MGDA [here](https://github.com/isl-org/MultiObjectiveOptimization/blob/d45eb262ec61c0dafecebfb69027ff6de280dbb3/multi_task/train_multi_task.py#L146). Please correct me if I'm wrong. Thanks!

**Comment by turboxin** [comment]
> I actually mean the lines [here](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L305) and [here](https://github.com/eniolasonowo/MotionNet/blob/ab664d594eef6e2aa88fd8f165c04086794413c1/train_single_seq_MGDA.py#L306), where the two optimizers are first zero-graded and then used to compute the grad. In mine understanding it seems that one optimizer could do the same thing, which is the case for origin implemantation of MGDA [here](https://github.com/isl-org/MultiObjectiveOptimization/blob/d45eb262ec61c0dafecebfb69027ff6de280dbb3/multi_task/train_multi_task.py#L146). Please correct me if I'm wrong. Thanks!

One optimizer means a optimizer for both shared feature extraction backbone and three heads, as the one [here](https://github.com/eniolasonowo/MotionNet/blob/ab664d594eef6e2aa88fd8f165c04086794413c1/train_multi_seq.py#L160)

**Comment by pxiangwu** [comment]
Hello,

Thank you for this question. The reason that I used two optimizers were due to the decoupled network structure of `MotionnetMGDA`, which consists of a feature encoder and heads. The reason I used such a decoupled network is due to these lines of code: [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L310), [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L331) and [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L356). Specifically, in the code
```
grads[0] = []
grads[0].append(shared_feats_tensor.grad.data.clone().detach())
shared_feats_tensor.grad.data.zero_()
...
grads[1] = []
grads[1].append(shared_feats_tensor.grad.data.clone().detach())
shared_feats_tensor.grad.data.zero_()
...
grads[2] = []
grads[2].append(shared_feats_tensor.grad.data.clone().detach())
shared_feats_tensor.grad.data.zero_()
```
I need to compute the gradients of a head with respect to the backbone **three** times. In other words, the same backbone grad data need to be reused **three** times. Therefore I particularly created a module `FeatEncoder` for this purpose.

Of course, we can have some other ways to get the gradients with respect to the shared backbone. At the time when I implemented the code, I simply used this somehow awkward strategy.

Hope my explanations help.

**Comment by pxiangwu** [comment]
Hello,

Thank you for this question. The reason that I used two optimizers were due to the decoupled network structure of `MotionnetMGDA`, which consists of a feature encoder and heads. The reason I used such a decoupled network is due to these lines of code: [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L310), [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L331) and [Link](https://github.com/eniolasonowo/MotionNet/blob/master/train_single_seq_MGDA.py#L356). Specifically, in the code
```
grads[0] = []
grads[0].append(shared_feats_tensor.grad.data.clone().detach())
shared_feats_tensor.grad.data.zero_()
...
grads[1] = []
grads[1].append(shared_feats_tensor.grad.data.clone().detach())
shared_feats_tensor.grad.data.zero_()
...
grads[2] = []
grads[2].append(shared_feats_tensor.grad.data.clone().detach())
shared_feats_tensor.grad.data.zero_()
```
I need to compute the gradients of a head with respect to the backbone **three** times. In other words, the same backbone grad data need to be reused **three** times. Therefore I particularly created a module `FeatEncoder` for this purpose.

Of course, we can have some other ways to get the gradients with respect to the shared backbone. At the time when I implemented the code, I simply used this somehow awkward strategy.

Hope my explanations help.

**Comment by turboxin** [comment]
Thank you for your detailed explanation! I understand now.
